Roadbase and ground spawn support

MetalStormGhost opened the pull request:
Implemented support for roadbases and ground spawn slots at airfields and FOBs. The ground spawn slots can be inserted in campaigns by placing either an A-10A or an AJS37 at a runway or ramp. This will cause an invisible FARP, an ammo dump and a fuel dump to be placed (behind the slot in case of A-10A, to the side in case of AJS37) in the generated campaigns. The ground spawn slot can be used by human controlled aircraft in generated missions. Also allowed the use of the four-slot FARP and the single helipad in campaigns, in addition to the invisible FARP. The first waypoint of the placed aircraft will be the center of a Remove Statics trigger zone (which might or might not work in multiplayer due to a DCS limitation).

![image](https://github.com/dcs-retribution/dcs-retribution/assets/89945461/62b2c662-aa96-46f7-a9ef-56c60aa262ff)
![image](https://github.com/dcs-retribution/dcs-retribution/assets/89945461/ba0bc49d-79e0-4399-86e0-5afad29abdfb)

Also implemented three new options in settings:
 - AI fixed-wing aircraft can use roadbases / bases with only ground spawns
   - This setting will allow the AI to use the roadbases for flights and transfers. AI flights will air-start from these bases, since the AI in DCS is not currently able to take off from ground spawns.
 - Spawn trucks at ground spawns in airbases instead of FARP statics
 - Spawn trucks at ground spawns in roadbases instead of FARP statics
   - These settings will replace the FARP statics with refueler and ammo trucks at roadbases. Enabling them might have a negative performance impact.

MetalStormGhost committed:
Roadbase and ground spawn support

Implemented support for roadbases and ground spawn slots at airfields and FOBs. The ground spawn slots can be inserted in campaigns by placing either an A-10A or an AJS37 at a runway or ramp. This will cause an invisible FARP, an ammo dump and a fuel dump to be placed (behind the slot in case of A-10A, to the side in case of AJS37) in the generated campaigns. The ground spawn slot can be used by human controlled aircraft in generated missions. Also allowed the use of the four-slot FARP and the single helipad in campaigns, in addition to the invisible FARP. The first waypoint of the placed aircraft will be the center of a Remove Statics trigger zone (which might or might not work in multiplayer due to a DCS limitation).

Also implemented three new options in settings:
 - AI fixed-wing aircraft can use roadbases / bases with only ground spawns
   - This setting will allow the AI to use the roadbases for flights and transfers. AI flights will air-start from these bases, since the AI in DCS is not currently able to take off from ground spawns.
 - Spawn trucks at ground spawns in airbases instead of FARP statics
 - Spawn trucks at ground spawns in roadbases instead of FARP statics
   - These settings will replace the FARP statics with refueler and ammo trucks at roadbases. Enabling them might have a negative performance impact.

MetalStormGhost committed:
Modified calculate_parking_slots() so it now takes into account also helicopter slots on FARPs and also ground start slots (but only if the aircraft is flyable or the "AI fixed-wing aircraft can use roadbases / bases with only ground spawns" option is enabled in settings).

MetalStormGhost committed:
Improved the way parking slots are communicated on the basemenu window.

Raffson commented on game/missiongenerator/aircraft/aircraftgenerator.py:
```diff
@@ -209,13 +226,6 @@ def create_and_configure_flight(
-            wpt.helipad_id = hpad.id
-            wpt.link_unit = hpad.id
-            self.helipads[flight.arrival].units.append(hpad)
-
```
Where is this being handled now?

MetalStormGhost replied:
Same place. It's now been restored.

MetalStormGhost committed:
Restored Set FREQ in tgogenerator.py, it was previously erroneously deleted.

Raffson commented on game/campaignloader/mizcampaignloader.py:
```diff
@@ -425,7 +442,66 @@ def add_preset_locations(self) -> None:
+                    PointWithHeading.from_point(
+                        static.position, Heading.from_degrees(static.units[0].heading)
+                    )
+                )
```
Refactor to a (private) method `add_helipad`?

MetalStormGhost replied:
Done.

Raffson commented on game/campaignloader/mizcampaignloader.py:
```diff
@@ -425,7 +442,66 @@ def add_preset_locations(self) -> None:
+                    ),
+                    first_waypoint,
+                )
+            )
```
Refactor to `add_roadbase` ?

MetalStormGhost replied:
Done.

Raffson commented on game/campaignloader/mizcampaignloader.py:
```diff
@@ -425,7 +442,66 @@ def add_preset_locations(self) -> None:
+                    ),
+                    first_waypoint,
+                )
+            )
```
Refactor to `add_roadbase_parking_slot` ?

MetalStormGhost replied:
Done.

Raffson commented on game/missiongenerator/aircraft/aircraftgenerator.py:
```diff
@@ -66,6 +69,8 @@ def __init__(
         self.flights: List[FlightData] = []
         self.mission_data = mission_data
         self.helipads = helipads
+        self.ground_spawns_roadbase = ground_spawns_roadbase
+        self.ground_spawns = ground_spawns
```
Would it make sense to refactor this into a container class? Something like `CustomAirport` ?

MetalStormGhost replied:
Maybe, but I don't want to do it here at this stage. The ground spawns are handled identically to helipads right now, but I'm not sure if that container class should only contain helipads and ground spawns?

Raffson replied:
That would make sense but yeah, if you want to keep this for later I'm all good with that 👍

Raffson commented on game/missiongenerator/aircraft/flightgroupspawner.py:
```diff
@@ -47,13 +47,17 @@ def __init__(
         self.flight = flight
         self.country = country
         self.mission = mission
         self.helipads = helipads
+        self.ground_spawns_roadbase = ground_spawns_roadbase
+        self.ground_spawns = ground_spawns
```
Looks like a refactor would definitely be a good idea here 😆

Raffson commented on game/missiongenerator/aircraft/flightgroupspawner.py:
```diff
@@ -283,6 +375,12 @@ def dcs_start_type(self) -> DcsStartType:
 
+    def _start_type_at_airfield(
+        self,
+        airfield: Airfield,
```
Unused?

MetalStormGhost replied:
It's used here:

>     def _generate_at_airfield(self, name: str, airfield: Airfield) -> FlyingGroup[Any]:
>         # TODO: Delayed runway starts should be converted to air starts for multiplayer.
>         # Runway starts do not work with late activated aircraft in multiplayer. Instead
>         # of spawning on the runway the aircraft will spawn on the taxiway, potentially
>         # somewhere that they don't fit anyway. We should either upgrade these to air
>         # starts or (less likely) downgrade to warm starts to avoid the issue when the
>         # player is generating the mission for multiplayer (which would need a new
>         # option).
>         self.flight.unit_type.dcs_unit_type.load_payloads()
>         return self.mission.flight_group_from_airport(
>             country=self.country,
>             name=name,
>             aircraft_type=self.flight.unit_type.dcs_unit_type,
>             airport=airfield.airport,
>             maintask=None,
>             start_type=self._start_type_at_airfield(airfield),
>             group_size=self.flight.count,
>             parking_slots=None,
>         )

Raffson commented on game/missiongenerator/missiongenerator.py:
```diff
@@ -226,6 +226,8 @@ def generate_air_units(self, tgo_generator: TgoGenerator) -> None:
             self.unit_map,
             mission_data=self.mission_data,
             helipads=tgo_generator.helipads,
+            ground_spawns_roadbase=tgo_generator.ground_spawns_roadbase,
+            ground_spawns=tgo_generator.ground_spawns,
```
More lines that would be saved by the refactor 😂

Raffson commented on game/missiongenerator/tgogenerator.py:
```diff
@@ -9,13 +9,107 @@
+from dcs import Mission, Point, unitgroup
 from dcs.action import DoScript, SceneryDestructionZone
 from dcs.condition import MapObjectIsDead
+from dcs.countries import (
```
`from dcs.countries import *`perhaps?

MetalStormGhost replied:
Done.

Raffson commented on game/game.py:
```diff
@@ -90,6 +90,8 @@ class TurnState(Enum):
 
 
 class Game:
+    scenery_clear_zones: List[Point]
```
I guess this can also be included in the changelog, since campaign designers will have the ability in general... though keep in mind to mention that it doesn't work properly in MP

MetalStormGhost replied:
Yeah. Good point. This has now been added to the changelog.

Raffson commented on game/procurement.py:
```diff
@@ -232,11 +237,15 @@ def best_squadrons_for(
             if not squadron.has_aircraft_capacity_for(request.number):
                 continue
-            if squadron.location.unclaimed_parking() < request.number:
+            if squadron.location.unclaimed_parking(parking_type) < request.number:
```
I this second check intentional?

MetalStormGhost replied:
Yes. has_aircraft_capacity_for() doesn't take into account the parking type, it only cares about the number. unclaimed_parking(), on the other hand, takes the parking type into account.

However, should I refactor has_aircraft_capacity_for() to actually care about the parking type?

Raffson replied:
Yeah it looks quite similar but I wouldn't refactor unless has_aircraft_capacity_for is only used here, but I doubt that's the case 😅

Raffson commented on game/theater/controlpoint.py:
```diff
@@ -757,17 +812,22 @@ def aircraft_retreat_destination(
     ) -> Optional[ControlPoint]:
         closest = ObjectiveDistanceCache.get_closest_airfields(self)
-        max_retreat_distance = squadron.aircraft.max_mission_range
+        # Multiply the max mission range by two when evaluating retreats,
+        # since you only need to fly one way in that case
+        max_retreat_distance = squadron.aircraft.max_mission_range * 2
```
Add "Improved squadron retreat logic" to changelog?

MetalStormGhost replied:
Added.

Raffson commented on qt_ui/windows/basemenu/airfield/QAircraftRecruitmentMenu.py:
```diff
@@ -91,8 +91,13 @@ def __init__(self, game_model: GameModel, control_point: ControlPoint) -> None:
     def update_label(self) -> None:
-        next_turn = self.control_point.allocated_aircraft()
-        max_amount = self.control_point.total_aircraft_parking
+        parking_type = ParkingType()
+        parking_type.include_rotary_wing = True
+        parking_type.include_fixed_wing = True
+        parking_type.include_fixed_wing_stol = True
```
I see this part returning a lot... perhaps there's a cleaner way to initialize this with the constructor?

MetalStormGhost replied:
There actually already was, I was just not using the contructor for that. Personal preference, I guess. I have now changed it as you suggest, and it does look cleaner.

Raffson committed:
Merge branch 'dev' into roadbase-support

Raffson committed:
Merge branch 'dev' into roadbase-support

Raffson committed:
Merge branch 'dev' into roadbase-support

Raffson committed:
Fix bad merge

MetalStormGhost committed:
Refactored helipad and ground spawn appends to static methods _add_helipad and _add_ground_spawn in mizcampaignloader.py
Added missing changelog entries.
Fixed tgogenerator.py imports.
Cleaned up ParkingType() construction.

MetalStormGhost committed:
Added missing plane_group argument to _add_ground_spawn calls.

MetalStormGhost committed:
Restored helicopter LANDING waypoint handling in aircraftgenerator.py

MetalStormGhost committed:
Added test_control_point_parking for testing that the correct number of parking slots are returned for control point in test_controlpoint.py

MetalStormGhost committed:
Cleaned up control_point construction in test_controlpoint.py

MetalStormGhost committed:
Added test_parking_type_from_squadron to test the correct ParkingType object is returned for a squadron of Viggens in test_controlpoint.py

MetalStormGhost committed:
Added test_parking_type_from_squadron to test the correct ParkingType object is returned for a squadron of Viggens in test_controlpoint.py

MetalStormGhost committed:
Added test_parking_type_from_squadron to test the correct ParkingType object is returned for a squadron of Viggens in test_controlpoint.py

MetalStormGhost committed:
Added test_parking_type_from_aircraft to test the correct ParkingType object is returned for Viggen aircraft type in test_controlpoint.py

MetalStormGhost committed:
Removed redundant if squadron.location.unclaimed_parking(parking_type) < request.number: check.
